gh-136003: Execute pre-finalization callbacks in a loop

ZeroIntensity opened the pull request:
<!-- gh-issue-number: gh-136003 -->
* Issue: gh-136003
<!-- /gh-issue-number -->

ZeroIntensity committed:
Execute early finalization handlers in a loop

ZeroIntensity committed:
Store Py_MAX result in a variable.

It re-evaluates it, apparently.

ZeroIntensity committed:
Add a test in test_atexit.

ZeroIntensity committed:
Deal with it in Py_EndInterpreter() as well.

ZeroIntensity committed:
Add a blurb entry.

ZeroIntensity committed:
Check the return code in the test.

ZeroIntensity committed:
Add a test for pending calls.

ZeroIntensity committed:
Fix tests on Windows.

sobolevn commented on Lib/test/test_atexit.py:
```diff
@@ -79,6 +79,30 @@ def thready():
+
+        return_code, stdout, stderr = script_helper.assert_python_ok("-c", textwrap.dedent(source))
+        self.assertEqual(return_code, 0)
+        end = "\r\n" if os.name == "nt" else "\n"
```
`os.linesep`?

ZeroIntensity replied:
I forgot that existed, thanks!

vstinner commented on Python/pylifecycle.c:
```diff
@@ -2003,6 +2003,42 @@ resolve_final_tstate(_PyRuntimeState *runtime)
+    /* Each of these functions can start one another, e.g. a pending call
+     * could start a thread or vice versa. To ensure that we properly clean
+     * call everything, we run these in a loop until none of them run anything. */
+    for (;;) {
```
Would it make sense to add an arbitrary limit to detect infinite loop? For example, log an error after 16 attemps.

ZeroIntensity replied:
Well, it would prevent deadlocks in rare cases, but it would cause crashes in other equally rare cases. Maybe it would be better to emit a *fatal* error when there are too many iterations?

ZeroIntensity committed:
Use os.linesep.

ericsnowcurrently commented on Lib/test/test_atexit.py:
```diff
@@ -79,6 +79,29 @@ def thready():
         # want them to affect the rest of the tests.
         script_helper.assert_python_ok("-c", textwrap.dedent(source))
 
+    @threading_helper.requires_working_threading()
+    def test_thread_created_in_atexit(self):
```
It would probably make sense to extend this with a subtest to check it in a subinterpreter.

ericsnowcurrently commented on Lib/test/test_capi/test_misc.py:
```diff
@@ -1641,6 +1642,36 @@ def subthread():
 
             self.assertEqual(actual, int(interpid))
 
+    @threading_helper.requires_working_threading()
+    def test_pending_call_creates_thread(self):
```
Likewise, it would probably make sense to extend this with a subtest to check it in a subinterpreter.

Also, it would be good to test adding a pending call to a different interpreter during its finalization.

ericsnowcurrently replied:
Did this get addressed?

ZeroIntensity replied:
No, because we can't use `_testcapi` in subinterpreters. Created #139102 to fix this.

ZeroIntensity replied:
Fixed.

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -2003,6 +2003,42 @@ resolve_final_tstate(_PyRuntimeState *runtime)
+        if (called == 0) {
+            break;
+        }
+    }
```
There's a race with this loop.  For example, a thread could be started by something other than an atexit handler or a pending call (i.e. from a daemon thread), and that could happen after `wait_for_thread_shutdown()` is called.  Likewise for the others.

I imagine we would need to do a check at the bottom of the loop and wrap it in a lock.  That means something like this:

```suggestion
        // Wrap up existing "threading"-module-created, non-daemon threads.
        wait_for_thread_shutdown(tstate);

        // Make any remaining pending calls.
        _Py_FinishPendingCalls(tstate);

        /* The interpreter is still entirely intact at this point, and the
        * exit funcs may be relying on that.  In particular, if some thread
        * or exit func is still waiting to do an import, the import machinery
        * expects Py_IsInitialized() to return true.  So don't say the
        * runtime is uninitialized until after the exit funcs have run.
        * Note that Threading.py uses an exit func to do a join on all the
        * threads created thru it, so this also protects pending imports in
        * the threads created via Threading.
        */

        _PyAtExit_Call(tstate->interp);
        
        // Check if we're done.
        PyMutex_Lock(tstate->interp->prefini_mutex);
        if (!count_nondaemon_threads(tstate)
            && !count_pending_calls(tstate)
            && !count_atexit_handlers(tstate))
        {
            // There's nothing left to clean up.  We keep the lock held
            // to prevent adding new cleanup items.  We will release it
            // after _PyInterpreterState_SetFinalizing() is called.
            break;
        }
        PyMutex_Unlock(tstate->interp->prefini_mutex);
```

Then we'd need to make sure the prefini lock keeps new threads from getting started and new pending calls and atexit handlers from getting added.

Note that we would release the prefini lock either right after we call `_PyInterpreterState_SetFinalizing()` or even *in* `_PyInterpreterState_SetFinalizing().

ZeroIntensity replied:
I think we probably can use an RW lock for this. I'll see what I can put together. Thanks for catching this!

ZeroIntensity replied:
Ok, I realized we didn't even need an extra lock: we can just use the GIL, and stop-the-world under FT. I still switched to your approach of counting rather than checking if they created callbacks, but we're sure that it's thread-safe because we either hold the GIL or stopped all other threads. Let me know what you think!

ericsnowcurrently reviewed (changes requested):
Overall, good work. 😄

I'm pretty sure there's a race we need to deal with which will require a variety of changes to this PR.

ZeroIntensity committed:
Use an RW lock instead of a counter.

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into fix-circular-finalization

ZeroIntensity committed:
Remove more counters.

ZeroIntensity committed:
Remove old artifacts (again).

ZeroIntensity committed:
Final time removing artifacts.

ZeroIntensity committed:
(I lied)

ZeroIntensity committed:
Atomically check if there are threads, atexit callbacks, or pending calls.

ZeroIntensity committed:
Remove stray newline change.

ZeroIntensity committed:
Add a test for atexit with subinterpreters.

ZeroIntensity committed:
Check for os.pipe() in the test.

ZeroIntensity committed:
Rely on stop-the-world and the GIL instead of a dedicated RW mutex.

This solution is much simpler, but will perform a little bit worse.
Instead of using a dedicated lock on the interpreter state, we can
simply use stop-the-world (and the GIL on the default build) to ensure that
no other threads can create pre-finalization callbacks concurrently.

ZeroIntensity committed:
Serialize pending calls via the ceval mutex.

Py_AddPendingCall() can be called without an attached thread state, so
we can't rely on stop-the-world or the GIL for it. Instead, acquire its
dedicated mutex to prevent threads from creating new pending calls.

ZeroIntensity committed:
Only check for non-daemon threads at finalization.

Since threading._shutdown() only shuts down non-daemon threads, daemon
threads would cause the loop to run indefinitely.

ZeroIntensity committed:
Fix assertion failures on the GILful build.

There were assertions that required the world to be stopped, but since
that isn't a thing on GILful builds, it caused failures. Protect those
assertions with a macro.

ZeroIntensity committed:
Merge branch 'main' into fix-circular-finalization

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -2011,6 +2011,105 @@ resolve_final_tstate(_PyRuntimeState *runtime)
+        // XXX Why does _PyThreadState_DeleteList() rely on all interpreters
+        // being stopped?
+        PyMutex_Lock(&interp->ceval.pending.mutex);
+        _PyEval_StopTheWorldAll(interp->runtime);
```
I forget, does this also cause threads created after this call to pause?

ZeroIntensity replied:
Yeah, no thread can attach while the world is stopped.

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -2011,6 +2011,105 @@ resolve_final_tstate(_PyRuntimeState *runtime)
+    for (;;) {
+        assert(!interp->runtime->stoptheworld.world_stopped);
+
+        // Wrap up existing "threading"-module-created, non-daemon threads.
+        wait_for_thread_shutdown(tstate);
```
We don't necessarily need to worry about it here, but it would probably also be worth waiting here for the interpreter to not be "running main".

ZeroIntensity replied:
Let's do that in a follow-up (probably tomorrow).

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -2011,6 +2011,105 @@ resolve_final_tstate(_PyRuntimeState *runtime)
+
+        // XXX Why does _PyThreadState_DeleteList() rely on all interpreters
+        // being stopped?
+        PyMutex_Lock(&interp->ceval.pending.mutex);
```
It may be worth preventing new pending calls via "is finalizing" and/or "world is stopped".  That doesn't need to be done right now, but at least there should be a comment here (and repeated at the new line at 2168) indicating *why* this lock is involved here.

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -2027,23 +2126,7 @@ _Py_Finalize(_PyRuntimeState *runtime)
-     */
-
-    _PyAtExit_Call(tstate->interp);
+    make_pre_finalization_calls(tstate);
```
For clarity, I'd add a comment here:

```
    // This call stops the world and takes the pending calls lock.
```

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -2441,13 +2525,7 @@ Py_EndInterpreter(PyThreadState *tstate)
-    _Py_FinishPendingCalls(tstate);
-
-    _PyAtExit_Call(tstate->interp);
+    make_pre_finalization_calls(tstate);
```
Likewise, I'd add a comment here:

```
    // This call stops the world and takes the pending calls lock.
```

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -2011,6 +2011,105 @@ resolve_final_tstate(_PyRuntimeState *runtime)
+
+    // We don't have to worry about locking this because the
+    // world is stopped.
+    _Py_FOR_EACH_TSTATE_UNLOCKED(interp, tstate) {
+        if (tstate->_whence == _PyThreadState_WHENCE_THREADING) {
+            return 1;
+        }
+    }
```
Using `tstate->_whence` feels a little hacky.  Is there an explicit indicator for how many threads have been created by the threading module?  If not, there's a strong justification here for adding one.

ericsnowcurrently replied:
I mean a counter on `PyInterpreterState` or get the info from the threading module or add something like `PyThreadState.blocks_fini` or that sort of thing.

ericsnowcurrently replied:
That better approach doesn't need to block this PR, but we'd need to follow up with the better approach soon after.

ZeroIntensity replied:
Sure, I don't mind refactoring afterwards.

ericsnowcurrently reviewed (approved):
I have a few follow-up comments/questions.

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into fix-circular-finalization

ZeroIntensity committed:
Fix merge conflict artifact.

ZeroIntensity committed:
Improve comments.

ZeroIntensity committed:
Merge branch 'fix-circular-finalization' of https://github.com/zerointensity/cpython into fix-circular-finalization

ZeroIntensity committed:
Fix ordering of finalization calls.

ZeroIntensity committed:
Finalize subinterpreters as a pre-finalization check.

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into fix-circular-finalization

ZeroIntensity committed:
Test _PyEval_AddPendingCall() instead of Py_AddPendingCall()

Apparently, Py_AddPendingCall() schedules for the main interpreter.

ZeroIntensity committed:
Add a test for subinterpreters.
